### DISCUSSION TITLE
Inject log variable from within a plugin

### DISCUSSION THREAD

**Discussion by AviBueno** [fastify/fastify#2785] [Q&A] [answered]
Sounds like this should be trivial, but I'm unable to either accomplish the task or find info regarding it.

From within a custom-made plugin, I'd like to inject a data variable that will appear in all log lines of a request.
I tried installing a child log upon `onRequest` but evidently that's not working as expected (see minimal sample below).

Am I doing it correctly or not?
Any other ideas on how to achieve this goal?

* Note: In a way, this issue reminds issue #2663

```
import fastify from 'fastify';
import pino from 'pino';
const logger = pino();
const server = fastify({ logger });
server.register(async (fastify) => {
  fastify.addHook('onRequest', async (request, reply) => {
    request.log = reply.log = request.log.child({ answer: 42 });
  });
});
```

**Comment by kibertoad** [comment]
So to clarify, then you are calling request.log.info('something something') and your custom metadata "answer" is not logged?

**Reply by AviBueno** [reply]
Exactly.
I have "reqId" logged and I'd like my custom data to be logged similarly alongside.

**Reply by AviBueno** [reply]
To clarify a bit more: If I'll log something inside that hook, it WILL log my custom data, but outside that hook and for the rest of the request's lifecycle, the custom data will vanish as if it were never there.

**Comment by kibertoad** [comment]
@AviBueno Have you tried setting logger to null there to see if it is being reset by something? Quick check inside fastify source shows that log is only set during Request creation, but maybe there is something else in your app doing its own reset.

**Comment by kibertoad** [comment, marked as answer]
@AviBueno Check this out: https://www.fastify.io/docs/master/Encapsulation/
What might be happening is that you are only applying this hook to an isolated context that is only available within that context. Can you try adding hook on the parent fastify instance (`server`) instead?

**Comment by AviBueno** [comment]
Hey @kibertoad, thank you for replying.

Following your comment I realized that in my test app I mistakenly used the `server` instance instead of the `fastify` instance in order to install the inner route.
Now that it's fixed, the log is dumped correctly.

The weird thing is that in my real code, everything is set up correctly, yet I don't see the extra attributes being logged.
There's no code that resets the log so I'm still stuck, but since the test case works then there the issue is local to my code and not to fastify itself so we can resolve this issue.

Thanks again for your help!

**Comment by kibertoad** [comment]
@AviBueno Glad it was helpful! And if you'll still feel stuck after investigating it for a while, please feel free to ping here again.

**Reply by AviBueno** [reply]
Thanks man, found the culprit: I'm managing a custom context object on the request, which (among other things) has a `log` member for convenience, something like `req.ctx = {..., log: request.log, ...}`.
So the `onRequest` should actually be `req.log = req.ctx.log = reply.log = req.log.child(...)`

Thank again for setting me on the right path!
